### PR TITLE
id: add support for showing SELinux context (--context/-Z)

### PR DESCRIPTION
**Pull request by jhscheer**
Thanks to @koutheir 's brand new (published today) [selinux crate](https://crates.io/crates/selinux), I could add SELinux (`--context/-Z`) support to `uu_id` (related: #2075).

On my local machine this passes GNU's coreutil Testsuite (8.32) for "tests/id/context.sh". It should also pass "tests/id/no-context.sh", but that depends on `uu_ls -Z` being implemented and therefore fails at the moment.

However, there are drawbacks:
* GNU's Testsuite says: `context.sh: skipped test: this system lacks SELinux support`. So there won't be automatic testing because apparently the CICD docker images lack selinux support.
* The `selinux-sys` crate fails to build on the `ubuntu-latest` docker image. There's a discussion about this going on [here](https://github.com/uutils/coreutils/pull/2075#issuecomment-832094432).
* ~~I had to bump the MSRV to 1.53 because the selinux crate has a dependency for `arrayvec` which needs 1.51 and the selinux crate uses the `io::ErrorKind::Unsupported` feature which needs 1.53.~~
I don't know how crucial these dependencies are, maybe they can be lowered, @koutheir ?
Edit: the dependencies have changed, see [below](https://github.com/uutils/coreutils/pull/2438#issuecomment-864514071).

Once we know how to deal with these drawbacks, adding SELinux Context support to the remaining uutils, e.g. `uu_ls`, `uu_cp`, `uu_stat`, etc. should be just as easy as this one.

### CODE REVIEW

**Comment by koutheir** [comment]
> I had to bump the MSRV to 1.53 because the selinux crate has a dependency for arrayvec which needs 1.51 and the selinux crate uses the io::ErrorKind::Unsupported feature which needs 1.53.
> I don't know how crucial these dependencies are, maybe they can be lowered, @koutheir ?

The `arrayvec` dependency is used to avoid unnecessary heap allocations, so it is an optimization.

The `std::io::ErrorKind::Unsupported` is used because I always use the latest version of Rust, and mainly because I don't actively try to reduce the MSRV.

I can remove these limitations if they are problematic.

**Comment by koutheir** [comment]
@jhscheer, I just removed the limitations stated above from the `selinux` crate. Please update to version `0.1.1` of the crate.

**Comment by jhscheer** [comment]
@koutheir I just noticed that "arrayvec" version = "0.5.2" is already in Cargo.toml.
So if the dependency `^0.7` could be lowered to `^0.5` there's no need to do without it.

**Comment by koutheir** [comment]
Between using old code, and not using code, I prefer to not use code.

**Comment by jhscheer** [comment]
I tried `selinux 0.1.1` and it looks like it can be compiled with the current MSRV `1.43.1` now.
However, compiling `selinux-sys 0.4.2` needs:
* [Traits on larger arrays](https://blog.rust-lang.org/2020/10/08/Rust-1.47.html#traits-on-larger-arrays) from rust `1.47.0`
* `std::slice::strip_prefix` from rust `1.51.0`

@koutheir I'm just documenting this here. I'm not implying you should go out of your way and change something.

Personally, I would prefer a MSRV bump (to at least 1.47) for this project, but I don't know all the implications that might have and also that's not for me to decide.

**Comment by tertsdiepraam** [comment]
> Personally, I would prefer a MSRV bump (to at least 1.47) for this project, but I don't know all the implications that might have and also that's not for me to decide.

If you have drivers for this change, you could open an issue for this. One potential problem with 1.47 is that it's not supported by the current stable Debian release. I'm not really sure whether that's an issue.

**Comment by sylvestre** [comment]
> One potential problem with 1.47 is that it's not supported by the current stable Debian release. I'm not really sure whether that's an issue.

it isn't ;)

**Comment by rivy** [comment]
@jhscheer , let's rebase this off of current master and see if it can be modified to give a clean/green CI with 1.51.

If so, we can push a release, increase the MSRV to 1.51, then merge this...

File: .github/workflows/CICD.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ env:
   PROJECT_NAME: coreutils
   PROJECT_DESC: "Core universal (cross-platform) utilities"
   PROJECT_AUTH: "uutils"
-  RUST_MIN_SRV: "1.43.1" ## v1.43.0
+  RUST_MIN_SRV: "1.51.0" ## v1.43.0
   RUST_COV_SRV: "2020-08-01" ## (~v1.47.0) supported rust version for code coverage; (date required/used by 'coverage') ## !maint: refactor when code coverage support is included in the stable channel
```
Comment (rivy) [inline .github/workflows/CICD.yml:15]:
```suggestion
  RUST_MIN_SRV: "1.51.0" ## v1.51.0
  RUST_COV_SRV: "1.51.0" ## supported rust version for code coverage; (date required/used by 'coverage') ## !maint: refactor when code coverage support is included in the stable channel
```

**Comment by sylvestre** [comment]
Could you please fix the conflicts? thanks

**Comment by jhscheer** [comment]
@rivy 
> @jhscheer , let's rebase this off of current master and see if it can be modified to give a clean/green CI with 1.51.
> 
> If so, we can push a release, increase the MSRV to 1.51, then merge this...

With 1.51 the `selinux` crate can be compiled.
However, there's still the much more grave problem that `selinux-sys` crate cannot be compiled in a docker image: 
```
thread 'main' panicked at 'selinux-sys: Failed to find 'selinux/selinux.h': Kind(NotFound)', /cargo/registry/src/github.com-1ecc6299db9ec823/selinux-sys-0.4.2/build.rs:222:10
```
This PR cannot be merged before we know how to tackle this problem.
There's a discussion going on about it [here](https://github.com/uutils/coreutils/pull/2075#issuecomment-832094432)

File: .github/workflows/CICD.yml [hunk 2]
```diff
@@ -13,8 +13,8 @@ env:
   PROJECT_NAME: coreutils
   PROJECT_DESC: "Core universal (cross-platform) utilities"
   PROJECT_AUTH: "uutils"
-  RUST_MIN_SRV: "1.43.1" ## v1.43.0
-  RUST_COV_SRV: "2020-08-01" ## (~v1.47.0) supported rust version for code coverage; (date required/used by 'coverage') ## !maint: refactor when code coverage support is included in the stable channel
+  RUST_MIN_SRV: "1.51.0" ## v1.51.0
+  RUST_COV_SRV: "1.51.0" ## supported rust version for code coverage; (date required/used by 'coverage') ## !maint: refactor when code coverage support is included in the stable channel
```
Comment (sylvestre) [inline .github/workflows/CICD.yml:17]: I think we expect a date here
if not, the comment needs to be updated

**Comment by tertsdiepraam** [comment]
This diff fixes the import in the tests:
```diff
diff --git a/Cargo.toml b/Cargo.toml
index d0e9ff2e..3499e14a 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -146,7 +146,7 @@ feat_os_unix_musl = [
 feat_require_crate_cpp = [
     "stdbuf",
 ]
-feat_selinux = ["id/selinux"]
+feat_selinux = ["id/selinux", "selinux"]
 # "feat_require_unix" == set of utilities requiring support which is only available on unix platforms (as of 2020-04-23)
 feat_require_unix = [
     "chgrp",
diff --git a/build.rs b/build.rs
index e9fe129e..4fbb27cc 100644
--- a/build.rs
+++ b/build.rs
@@ -28,8 +28,8 @@ pub fn main() {
         if val == "1" && key.starts_with(env_feature_prefix) {
             let krate = key[env_feature_prefix.len()..].to_lowercase();
             match krate.as_ref() {
-                "default" | "macos" | "unix" | "windows" => continue, // common/standard feature names
-                "nightly" | "test_unimplemented" => continue,         // crate-local custom features
+                "default" | "macos" | "unix" | "windows" | "selinux" => continue, // common/standard feature names
+                "nightly" | "test_unimplemented" => continue, // crate-local custom features
                 "test" => continue, // over-ridden with 'uu_test' to avoid collision with rust core crate 'test'
                 s if s.starts_with(feature_prefix) => continue, // crate feature sets
                 _ => {}             // util feature name
```
The problem is that `selinux` was not exposed to the tests so the dependency needs to be enabled (and then ignored in `build.rs`)

**Comment by koutheir** [comment]
The `selinux-sys` crate requires `libselinux` headers and shared library to be accessible in the C toolchain. The crate can be built inside a container if this condition it true. Building it inside a container that does not provide the headers causes the error `Failed to find 'selinux/selinux.h': Kind(NotFound)`. Furthermore, if a cross-compilation toolchain is used to cross-compile the `selinux-sys` crate, then the `libselinux` headers and shared library need to be provided by that toolchain (the host system headers and library are not used in this case). Using a cross-compilation toolchain that does not provide the `libselinux` headers results in the same issue.

A potential solution to this complication would be to ship `libselinux` sources with the `selinux-sys` crate, and cross-compile `libselinux` right before compiling `selinux-sys`. But in order to retain compatibility, the `selinux-sys` crate needs to know the exact version of `libselinux` to be used on the target. Using the same version of `libselinux` provided by the build system is not possible, because `libselinux` does not provide any means to detect its version, neither in the header files, nor in the shared library itself, unfortunately.

Using the `cross` tool to build `selinux-sys` works in some cases, but custom containers need to be built (based on the `cross` containers, with added packages) in order to install `libselinux` and `libclang` and friends. I don't think this is a sustainable solution.

**Comment by jhscheer** [comment]
This is ready for review.

With new feature flag, the SELinux specific parts are now gated behind `--featues feat_selinux` and this can  be compiled in the CI.

Since this is now decoupled from the CICD, I appreciate a volunteer to run:
`cargo test test_id --features feat_selinux` on an SELinux enabled Linux distribution with libselinux headers installed (e.g. on fedora `dnf install libselinux-devel`).

The feature works like this:
```bash
$ cargo run --features feat_selinux -- id -Z
unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
$ cargo run --features "id" -- id -Z
id: --context (-Z) works only on an SELinux-enabled kernel
```

**Review by tertsdiepraam — APPROVED** [review]

**Comment by jhscheer** [comment]
@koutheir could you change `selinux-sys` to not use `std::slice::strip_prefix`?
Then this PR could be merged with only a bumb to MSRV 1.47 instead of 1.51 (https://github.com/uutils/coreutils/pull/2075#issuecomment-890327017).

I believe this PR could also be what you were talking about here (https://github.com/uutils/coreutils/pull/2075#issuecomment-890335344):
> I'm waiting for any other PR to handle the issue of depending on SELinux and still please the CI servers somehow. Once that is done, I'll modify this PR to handle the issue in a similar fashion.

**Comment by koutheir** [comment]
> @koutheir could you change `selinux-sys` to not use `std::slice::strip_prefix`?
> Then this PR could be merged with only a bumb to MSRV 1.47 instead of 1.51 ([#2075 (comment)](https://github.com/uutils/coreutils/pull/2075#issuecomment-890327017)).

I don't understand why `1.47` is preferable to `1.51`, but, yes, I can do that.

> I believe this PR could also be what you were talking about here ([#2075 (comment)](https://github.com/uutils/coreutils/pull/2075#issuecomment-890335344)):
> 
> > I'm waiting for any other PR to handle the issue of depending on SELinux and still please the CI servers somehow. Once that is done, I'll modify this PR to handle the issue in a similar fashion.

I believe so too. Once that issue is addressed somehow, I'll modify PR #2075 in order to use the same approach if possible.

**Comment by koutheir** [comment]
> @koutheir could you change `selinux-sys` to not use `std::slice::strip_prefix`?
> Then this PR could be merged with only a bumb to MSRV 1.47 instead of 1.51 ([#2075 (comment)](https://github.com/uutils/coreutils/pull/2075#issuecomment-890327017)).

The requested change is implemented by version `0.5.1` of `selinux-sys`.

**Comment by jhscheer** [comment]
> > @koutheir could you change `selinux-sys` to not use `std::slice::strip_prefix`?
> > Then this PR could be merged with only a bumb to MSRV 1.47 instead of 1.51 ([#2075 (comment)](https://github.com/uutils/coreutils/pull/2075#issuecomment-890327017)).
> 
> The requested change is implemented by version `0.5.1` of `selinux-sys`.

Thanks!
Could you please also update `selinux` crate? It is also using `std::slice::strip_prefix`.
https://github.com/koutheir/selinux/blob/f51890d3c5530a691b7e8e7737275e1ec1f2894a/src/lib.rs#L1157
Sorry, I didn't see this before.

**Comment by koutheir** [comment]
> Could you please also update `selinux` crate? It is also using `std::slice::strip_prefix`.
> https://github.com/koutheir/selinux/blob/f51890d3c5530a691b7e8e7737275e1ec1f2894a/src/lib.rs#L1157

The requested change is implemented by version `0.1.3` of the `selinux` crate.

**Comment by sylvestre** [comment]
Well done :)

**Pull request merged**